walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,223 @@
+      ArrayList<TimeRange> mandatoryAttendeeMeetingTimes = new ArrayList<>();
+      for (int b = 0; b < mandatoryAttendeeMeetings.size(); b++) {
+        mandatoryAttendeeMeetingTimes.add(mandatoryAttendeeMeetings.get(b).getWhen());
+      }
+      Collections.sort(mandatoryAttendeeMeetingTimes, Comparator.comparingInt(TimeRange::start)); 
+
+      //check for nested or overlapping meetings
+      int j = 0;
+      while (j < mandatoryAttendeeMeetingTimes.size()-1) {
+        TimeRange currentMeetingTime = mandatoryAttendeeMeetingTimes.get(j);
+        TimeRange nextMeetingTime = mandatoryAttendeeMeetingTimes.get(j+1);
+        
+        if (currentMeetingTime.contains(nextMeetingTime)) {
+          mandatoryAttendeeMeetingTimes.remove(j+1);
+          continue;
+        } 
+        else if (currentMeetingTime.overlaps(nextMeetingTime)) {
+          TimeRange totalMeetingTime = TimeRange.fromStartEnd(currentMeetingTime.start(), nextMeetingTime.end(), false);
+          mandatoryAttendeeMeetingTimes.set(j, totalMeetingTime);
+          mandatoryAttendeeMeetingTimes.remove(j+1);
+          continue;
+        } 
+        else {
+          j++;
+        }
+      }
+
+      //if no conflicting meetings (all were removed), any time of day works  
+      if (mandatoryAttendeeMeetingTimes.size() == 0) {
+        mandatoryAttendeeAvailability.add(TimeRange.WHOLE_DAY);
+      }
+
+      //otherwise, find available time ranges outside of meetings 
+      else {
+        
+        //time before first meeting
+        TimeRange firstAvailableTime = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, mandatoryAttendeeMeetingTimes.get(0).start(), false);
+        if (firstAvailableTime.duration() > 0) {
+          mandatoryAttendeeAvailability.add(firstAvailableTime);
+        }
+
+        //times between meetings
+        if (mandatoryAttendeeMeetingTimes.size() > 1) {
+          for (int k = 1; k < mandatoryAttendeeMeetingTimes.size(); k++) {
+            TimeRange nextAvailableTime = TimeRange.fromStartEnd(mandatoryAttendeeMeetingTimes.get(k-1).end(), mandatoryAttendeeMeetingTimes.get(k).start(), false);
+            if (nextAvailableTime.duration() > 0) {
+              mandatoryAttendeeAvailability.add(nextAvailableTime);
+            }
+          }
+        }
+        
+        //time after last meeting
+        TimeRange lastAvailableTime = TimeRange.fromStartEnd(mandatoryAttendeeMeetingTimes.get(mandatoryAttendeeMeetingTimes.size()-1).end(), TimeRange.END_OF_DAY, true);
+        if (lastAvailableTime.duration() > 0) {
+          mandatoryAttendeeAvailability.add(lastAvailableTime);
+        }
+      }
+
+      //remove any time ranges that aren't long enough 
+      int m = 0;
+      while (m < mandatoryAttendeeAvailability.size()) {
+        if (mandatoryAttendeeAvailability.get(m).duration() < request.getDuration()) {
+          mandatoryAttendeeAvailability.remove(m);
+          continue;
+        } else {
+          m++;
+        }
+      }
```

Comment:
Instead of looping through `mandatoryAttendeeAvailability` here to remove time ranges that aren't long enough, you could check before you even add it to `mandatoryAttendeeAvailability` (like where you check that the duration > 0 above).